README.md
    Snapwell is free software: you can redistribute it and/or modify it under
    the terms of the GNU General Public License as published by the Free
    Software Foundation, either version 3 of the License, or (at your option)
    any later version.

    Snapwell is distributed in the hope that it will be useful, but WITHOUT ANY
    WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
    FOR A PARTICULAR PURPOSE.

    See the GNU General Public License at <http://www.gnu.org/licenses/gpl.html>
    for more details.

# Usage

Snapwell requires an installation of [ert](https://github.com/Ensembles/ert) to
function.  Ert provides the necessary library to read Eclipse grids and restart
files.

Make sure to have snapwell in your Python path:

    you@host:~$ git clone https://github.com/Statoil/snapwell
    you@host:~$ export PYTHONPATH=snapwell:$PYTHONPATH
    you@host:~$ python
    >>> import snapwell

# On snapwell

Snapwell is a project for depth optimizing wellpaths by changing the TVD
according to OWC.  What exactly defines OWC at a given (x,y)-coordinate can be
specified as number r between 0 and 1, and OWC is then the lowest point with
SWAT=r.  Future implementations might also take the permeability into account.

Changes in TVD are typically done by using a max increase of 0.5 m/30m to avoid
making vertical segments.

Should avoid putting the majority of the well path in cells with non producible
permeability.


# Fileformat
## snapwell input file

The config file format for snapwell (SnapConfig) is as follows:
* Everything on a line following a "--" is ignored
* A line should start with a keyword and be followed by the keyword's arguments
* Two keywords are mandatory, `GRID` and `RESTART`, and should be followed by a
  path to the grid (GRID/EGRID) and restart files (UNRST), respectively (see
  example below).
* The `INIT` keyword is required for permeability output, and should be followed
  by a path to an init file.
* The following additional header keywords are supported:
    * `OUTPUT` --- the output directory (path), is set to cwd by default
    * `OVERWRITE` --- a boolean (True/False) specifying whether the program is
      allowed to overwrite already existing files.  Defaulted to False; use with
      care.
    * `DELTA_Z` --- the maximum elevation per 100 meters, given as a float.
      0.0167 is 0.5m/30m.
    * `OWC` --- the saturation of water considered to be the OWC (e.g. 0.7)
    * `OWC_OFFSET` --- a number (eg. 0.5).  Will place the wellpoint 0.5m above
      OWC
    * `LOG` --- any number of LOG lines is allowed, and supported are:
       * `OWC` --- actual OWC for given date and (x,y) coordinate
       * `OLD_TVD` --- the old TVD specified in the input Wellpath
       * `DIFF` --- The difference between `OLD_TVD` and new `TVD`
       * `SWAT`, `SOIL`, `SGAS`, `PERMX` --- value in wellpoint cell for keyword
       * `LENGTH` --- the distance between this (x,y) coordinate and the previous
* then follow n lines containing the keyword `WELLPATH` followed by two or four
  literals
    * a filepath (path to the wellpath)
    * a datetime (format specified below)
    * optionally "MD <float>" or "TVD <float>".

The filepath is relative to the config file, not the current working directory.

A datetime is specified as either
* YYYY
* YYYY-MM
* YYYY-MM-DD
If either MM or MM-DD is missing, they will be treated as 01 or 01-01, respectively.

Example:

    GRID       NORNE_1996.EGRID
    RESTART    NORNE_1996.UNRST
    INIT       NORNE_1996.INIT
    --
    OUTPUT        ../snapout
    OVERWRITE     False
    DELTA_Z       0.0167 -- 0.5m per 30m
    OWC_OFFSET    0.8    -- places well 0.8m above OWC
    LOG           SWAT   -- outputs SWAT value for the wellpoint's cell
    LOG           OWC    -- outputs actual OWC for date and x,y coordinate
    --
    --KEYWORD  FILENAME                     DATE        D_TYPE  D_SPEC
    WELLPATH   case1/SIMULATION_WELL_D_p.txt  2022-01-01  MD      2000.0
    WELLPATH   case1/SIMULATION_WELL_D_q.txt  2022        TVD     1750.0
    WELLPATH   case1/SIMULATION_WELL_D_r.txt  2022-05
    WELLPATH   case1/SIMULATION_WELL_E_f.txt  2022        MD      2558.5
    WELLPATH   case1/SIMULATION_WELL_E_m.txt  2022        MD      2300.0
    WELLPATH   case1/SIMULATION_WELL_E_n.txt  2022-12-24






## WellPath format (RMS well format)

The wellpath format is given as follows: (blank lines and lines starting with -- ignored)

    <version>
    <well type>
    <well name> <x> <y> <rkb>
    <number of well logs>
    <log name 1> <unit> <scale>
    ...
    <log name n> <unit> <scale>
    <x> <y> <z> <p1> ... <pn>
    ...
    <x> <y> <z> <p1> ... <pn>

An example is the following:

    1.0
    PRODUCTION
    TEST_WELL 1068.000 0.000 0.0000
    3
    MD 1 lin
    Incl 1 lin
    Az 1 lin
    1068.0    0.0    38.0    38.0    109.0    359.0
    1069.0    5.0    39.0    50.0    109.0    359.0

Here, the X, Y, and Z numbers are UTM* easting, UTM northing, UTM TVD (true
vertical depth), and then follow the three log parameters defined, MD, INC and
AZ, being measured_depth, inclination and azimuth.

We safely ignore all but the three first numbers in each row (actually, only x
and y are really interesting).

* UTM is Universal Transverse Mercator coordinate system

## Run tests
[tox](https://tox.readthedocs.io/en/latest/) is used as the test facilitator,
to run the full test suite:

```sh
# Test
pip install tox
tox
```

or to run it for a particular Python version (in this case Python 3.7):

```sh
# Test
pip install tox
tox -e py37
```

or to run it for a the current Python version on PATH:

```sh
# Test
pip install tox
tox -e py
```

[pytest](https://docs.pytest.org/en/latest/) is used as the test runner, so for quicker
iteration it is possible to run:

```sh
# Test
pytest
```

this requires that the `pytest` is installed.

```sh
# Install pytest
pip install pytest
```

[pre-commit](https://pre-commit.com/) is used to comply with the formatting standards.
The complete formatting tests can be run with:

```sh
pip install tox
tox -e style
```

See `.pre-commit-config.yaml` for the complete steps.

[pre-commit](https://pre-commit.com/) can also provide git hooks to run on every commit
to avoid commiting with formatting errors. This will only run on the diff so is quite fast.
To configure this, run:

```sh
pip install pre-commit
pre-commit install
```

After this the hook will run on every commit.

If you would like to remove the hooks, run:

```sh
pre-commit uninstall
```


release-notes.md
# Version 1.0.0 (2016-09-16)

PO confirms snapwell is stable enough to be called 1.0.

This is version 0.9.1 with a new name.

# Version 0.9.1 (2016-09-12)

Output the correct MD value.

General code cleanup.

Updated documentation and this file.

MR: !382 !383

JIRA:RES-23, JIRA:RES-25

# Version 0.9 (2016-09-09)

This release should be considered as an alpha version of 1.0.  Once we get a
green light from PO, we will bump the version to snapwell 1.0, and call it
stable.

Improved command line usage with argparse.  Can now use

    snapwell --owc_definition SWAT:0.65 --owc_offset 0.15 --delta 0.0167 --output ~/myfolder ~/statoil/myconf.sc

where
 * `--owc_definition` (or `-f`) set to `SWAT:0.65` means that the definition of
   OWC is the lowest point where the SWAT value goes from _above_ 0.65 to
   _below_ 0.65.
 * `--owc_offset` (or `-z`) is the vertical offset (in meters) _above_ OWC we
   want to put the wellpoint.  Set to 0.15 means that we aim the wellpoint
   towards the first cell center above OWC - offset. (Minus means up.)
* `--delta` (or `-d`) means the dogleg restrictions (in m/m).  Set to 0.0167, it
  means we can differ in TVD 0.5 meters per 30 meters in length difference (see
  `LOG LENGTH`)

Adding `-w` means we overwrite existing files if necessary.  The flag `-r` means
output ResInsight compatible files.  In this case, we ignore all `LOG` keywords
provided and output only `WELLNAME` and then `X`, `Y`, `TVD`, and `MD` for each
wellpoint.

With that feature, one can run parallell instance with one config file, e.g.

    snapwell -z 0.0 -o ~/snap0 myconf.sc
    snapwell -z 0.2 -o ~/snap2 myconf.sc
    snapwell -z 0.5 -o ~/snap5 myconf.sc

which would run Snapwell with `owc_owffset=0.0, 0.2, 0.5`, respectively, and
output the files to different folders, `~/snap0`, `~/snap2`, and `~/snap5`,
respectively.

MR: !373 !380

JIRA:RES-18


# Version 0.8.3

Some bug fixes and improvements.

Update RKB values to be x,y,md-tvd for the first row (see issue #7)

Can now output columns SWAT, SGAS, SOIL (with LOG keyword).

MR: !375.

JIRA:RES-20, JIRA:RES-21


# Version 0.8.2

Wellpath aims towards first cell center above OWC+OWC_OFFSET, where OWC is
defined in config file (default SWAT 0.7).

If a user wants to have the first (lowest) "oil cell", one should use
OWC_OFFSET=0 in the config file.

JIRA:RES-16--RES-17


# Version 0.8.1

Bug fix for nan-values.

ert/ert-statoil:#6 https://git.statoil.no/ert/ert-statoil/issues/6


# Version 0.8.0 (2016-08-30)

Use LOG keywords to optionally add colums to output.
Supported keywords are:

    LOG LENGTH
    LOG TVD_DIFF
    LOG OLD_TVD
    LOG OWC
    LOG PERMX

JIRA:RES-11



# Version 0.7.4

Start searching for OWC at bottom.  Fixed interpolation algorithm.

JIRA:RES-14

# Version 0.7.3

Fixed bugs in snap algorithm, use 2D distance (not 3D) for calculating length.

# Version 0.7.2

Added WINDOW_DEPTH to config.  See readme for usage.

JIRA:RES-12

# Version 0.7.1

Removed cell center logic (viz meeting 2016-08-25 with Mathias)

Wells snap to thresh-owc_offset (owc=0.7 - 0.5m by default), unless we are close
to an even integer, in which case, we round about 0.1m away.

JIRA:RES-13.


# Version 0.7 (2016-08-23)
First trial version.  This version is a minimal working snapwell application.

Parsing wellpaths and config file.
* Understands x,y,z, can add arbitrary columns to wellpaths
* Parsing of wellpaths, dates
* Parsing of output config as well as overwrite flag
* Parsing of OWC, OWC_OFFSET, DELTA_Z keywords

Visualizing wellpaths (reads config files as well)

Contains snapping feature and a first attempt on delta limits.

Contains a preliminary WINDOW_DEPTH keyword parsing.

JIRA:RES-1--RES-8.


snapwell/snapconfig.py
#  Copyright (C) 2016  Statoil ASA, Norway.
#
#  This file is part of Snapwell.
#
#  Snapwell is free software: you can redistribute it and/or modify it under the
#  terms of the GNU General Public License as published by the Free Software
#  Foundation, either version 3 of the License, or (at your option) any later
#  version.
#
#  Snapwell is distributed in the hope that it will be useful, but WITHOUT ANY
#  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
#  A PARTICULAR PURPOSE.
#
#  See the GNU General Public License at <http://www.gnu.org/licenses/gpl.html>
#  for more details.

from wellpath import WellPath
from snap_utils import read_config, parse_date, Inf, Nan
from ert.config import ContentTypeEnum, ConfigParser

from ert.ecl import EclGrid, EclFile

import datetime
from os import path, makedirs
import logging
from math import isnan

class SnapConfig:
    """A class for representing a Snapwell config object.  It contains
       the filename (path) for grid, for restart file, and a list of
       (welldata file, date)-pairs.
    """

    def __init__(self, grid, restart, init = None):
        self._grid = None
        self._restart = None
        self._init = None
        self._gridFile = grid
        self._restartFile = restart
        self._initFile = init
        self._wellpath_config = []
        self._output = '.'              # Output path
        self._overwrite = False         # Are we allowed to overwrite files?
        self._delta = Inf               # max ascend/descend rate for wellpaths
        self._owc_offset = 0.5          # default OWC Z offset
        self._owc_definition = ('SWAT', 0.7) # definition of OWC
        self._logKeywords = []

    def __len__(self):
        return len(self._wellpath_config)

    def __getitem__(self, index):
        return self._wellpath_config[index]

    def __str__(self):
        m = {}
        m["grid"] = self._gridFile
        m["restart"] = self._restartFile
        m["init"] = self._initFile
        m["wellpath"] = self._wellpath_config
        return "snap: " + str(m)

    def append(self, elt):
        """Append an elt=(fname, date) pair to wellpath.  In this setting, elt[0] should
           be a string that is a filepath to a wellpath and elt[1] should be
           datetime, corresponding to the restart time of this wellpath.

        """
        if type(elt[1]) is not type(datetime.datetime(2000,1,1)):
            raise TypeError("Need datetime object, not " + str(elt[1]))
        self._wellpath_config.append(elt)

    def setOwcOffset(self, owc):
        """This is the OWC Z offset value used.  That is, if OWC in the GRID is found to
           be at 1761.3, and owc=0.5 (the usual value), then we are interested
           in placing the wellpoint at 1761.3-0.5=1760.8.

        """
        self._owc_offset = float(owc)

    def owcOffset(self):
        return self._owc_offset


    def addLogKeyword(self, log):
        """ Add log keywords items
        """
        self._logKeywords.append(log)

    def logKeywords(self):
        return self._logKeywords


    def owcDefinition(self):
        """The definition of OWC might be owcDefinition=('SWAT', 0.7).  This
           means that OWC is the first z-point from the bottom that has cell
           property SWAT=0.7 (interpolated).  Can also be SGAS=0.1 ..."""
        return self._owc_definition

    def setOwcDefinition(self, owc_definition):
        if len(owc_definition) != 2:
            raise ValueError('OWC Definition must be a (str, float)-pair, not %s' % str(owc_definition))
        self._owc_definition = tuple(owc_definition)

    def setDeltaZ(self, delta):
        """This delta is the max inclination of a wellpath, i.e., we cannot have h_diff
        (horizontal difference) between two wellpoints h_diff/z_diff > delta.
        """
        self._delta = float(delta)

    def deltaZ(self):
        return self._delta

    def setOutput(self, path):
        """The output folder to which we write the new wellpaths."""
        self._output = path

    def output(self):
        return self._output

    def setOverwrite(self, overwrite):
        """If this is set to True, we will overwrite the wellpath files.

           This should not be set to true unless you know what you are doing,
           hence the rigidity in "if overwrite is True".
        """
        self._overwrite = False
        if overwrite is True:
            self._overwrite = True

    def overwrite(self):
        return self._overwrite is True

    def getWellpath(self, idx):
        fname = self.filename(idx)
        wp = WellPath.parse(fname)
        if wp.wellname() and len(wp.wellname()) > 1 and len(wp.wellname().split()) == 1:
            wp.setFilename(path.abspath(path.join(self._output, wp.wellname())))
        if self.depthType(idx):
            wp.setDepthType(self.depthType(idx))
            wp.setWindowDepth(self.windowDepth(idx))
            print("Configuring depth: %s [%s]" % (str(wp.windowDepth()),str(wp.depthType())))
        return wp

    def filename(self, idx):
        """Return filename of idx'th wellpath."""
        return self._wellpath_config[idx][0]
    
    def date(self, idx):
        """Return date of idx'th wellpath."""
        return self._wellpath_config[idx][1]

    def depthType(self, idx):
        """Return depth type (MD/TVD) of idx'th wellpath or None if not set."""
        if len(self._wellpath_config[idx]) == 4:
            return self._wellpath_config[idx][2]
        return None

    def windowDepth(self, idx):
        """Return depth of idx'th wellpath or -inf if not set."""
        if len(self._wellpath_config[idx]) == 4:
            return self._wellpath_config[idx][3]
        return -Inf

    def gridFile(self):
        return self._gridFile
    def restartFile(self):
        return self._restartFile
    def initFile(self):
        return self._initFile

    def getGrid(self):
        if self._grid is None:
            self._grid = EclGrid(self.gridFile())
        return self._grid

    def getRestart(self):
        if self._restart is None:
            self._restart = EclFile(self.restartFile())
        return self._restart

    def getInit(self):
        if self._init is None and self._initFile:
            self._init = EclFile(self.initFile())
        return self._init


    @staticmethod
    def tryset(setter, value, content):
        if not value in content:
            return False
        try:
            val = content[value][0][0]
            setter(val)
            return True
        except Exception as err:
            logging.warning('Ill specified %s.  Ignoring. %s' % (value, err))

    @staticmethod
    def addConfigItemFloat(parser, value, repeated=False):
        item = parser.add(value, repeated)
        item.iset_type(0, ContentTypeEnum.CONFIG_FLOAT)

    @staticmethod
    def addConfigItemPath(parser, value, repeated=False):
        item = parser.add(value, repeated)
        item.iset_type(0, ContentTypeEnum.CONFIG_PATH)


    @staticmethod
    def parse(fname):
        """Takes a configuration object containing grid file path, restart file
           path and a list of wellpath-files and their restart dates.  Returns a
           SnapWell object, containing this info.

        """

        conf = ConfigParser()


        SnapConfig.addConfigItemPath(conf, 'GRID')
        SnapConfig.addConfigItemPath(conf, 'RESTART')
        SnapConfig.addConfigItemPath(conf, 'INIT')
        SnapConfig.addConfigItemPath(conf, 'OUTPUT')

        overwrite_item = conf.add("OVERWRITE", False) # the last OVERWRITE specified counts
        overwrite_item.iset_type(0 , ContentTypeEnum.CONFIG_STRING )

        SnapConfig.addConfigItemFloat(conf, 'OWC_OFFSET')
        SnapConfig.addConfigItemFloat(conf, 'DELTA_Z')

        owc_def_item = conf.add('OWC_DEFINITION') # e.g. OWC_DEFINITION SWAT 0.7
        owc_def_item.iset_type(0 , ContentTypeEnum.CONFIG_STRING )
        owc_def_item.iset_type(1 , ContentTypeEnum.CONFIG_FLOAT )


        log_item = conf.add('LOG')
        log_item.iset_type(0 , ContentTypeEnum.CONFIG_STRING )

        wellpath_item = conf.add("WELLPATH", True) # a series of WELLPATH/DATE pairs
        wellpath_item.iset_type(0 , ContentTypeEnum.CONFIG_PATH )
        wellpath_item.iset_type(1 , ContentTypeEnum.CONFIG_STRING )

        content = conf.parse(fname)

        # Grid
        gridFile    = tryGetPath(content, 'GRID', 0, 0)
        if not gridFile:
            logging.info('No GRID file specified?')
            raise ValueError('Could not load GRID file from Snapwell config file.')

        # Restart
        restartFile = tryGetPath(content, 'RESTART', 0, 0)
        if not restartFile:
            logging.info('No RESTART file specified?')
            raise ValueError('Could not load RESTART file from Snapwell config file.')

        # init
        initFile = None
        if "INIT" in content:
            initFile = tryGetPath(content, 'INIT', 0, 0)
        s = SnapConfig(gridFile, restartFile, initFile)

        # output
        outputPath = None
        if "OUTPUT" in content:
            outputPath = tryGetPath(content, 'OUTPUT', 0, 0)
            if path.exists(outputPath):
                if path.isfile(outputPath):
                    raise ValueError('Provided output folder is a file.  Either delete file, or set a different output folder: %s'
                                     % outputPath)
                else:
                    s.setOutput(outputPath)
            else:
                makedirs(outputPath)
                s.setOutput(outputPath)

        # overwrite
        overwrite = False
        if 'OVERWRITE' in content:
            overwrite = tryGet(content, 'OVERWRITE', 0, 0)
            try:
                if overwrite.strip().lower() == 'true':
                    overwrite = True
            except Exception as err:
                logging.warning('Ill specified overwrite flag: "%s".' % err)
        s.setOverwrite(overwrite)


        SnapConfig.tryset(s.setDeltaZ,      'DELTA_Z',      content)
        SnapConfig.tryset(s.setOwcOffset,   'OWC_OFFSET',   content)


        # Loading OWC_DEFINITION
        if 'OWC_DEFINITION' in content:
            try:
                if len(content['OWC_DEFINITION'][0]) != 2:
                    logging.warning('Wrong number of arguments in OWC_DEFINITION.  Needs to be e.g. SWAT 0.7, got %s'
                                    % str(content['OWC_DEFINITION'][0]))
                else:
                    owc_kw  = tryGet(content, 'OWC_DEFINITION', 0, 0)
                    owc_val = tryGetFloat(content, 'OWC_DEFINITION', 0, 1)
                    s.setOwcDefinition((owc_kw, owc_val))
            except Exception as err:
                logging.warning('Ill specified OWC_DEFINITION keyword: "%s".' % err)
        else:
            logging.info('Using OWC definition %s' % str(s.owcDefinition()))


        if 'LOG' in content:
            for i in range(len(content['LOG'])):
                line = content['LOG'][i]
                num_tokens = len(line)
                if num_tokens < 1:
                    raise ValueError('Missing data in LOG %d: "%s".' % (i+1, str(line)))
                s.addLogKeyword(tryGet( content, 'LOG', i, 0))

        # Loading the wellpath file names
        if 'WELLPATH' not in content:
            logging.warning('No wellpaths provided in Snapwell config file.')
            return s

        for i in range(len(content['WELLPATH'])):
            wp_line = content['WELLPATH'][i]
            num_tokens = len(wp_line)
            if num_tokens < 2:
                raise ValueError('Missing data in WELLPATH %d: "%s".' % (i+1, str(wp_line)))

            fname    = tryGetPath(content, 'WELLPATH', i, 0)
            date_str = tryGet( content, 'WELLPATH', i, 1)

            if num_tokens > 2 and num_tokens != 4:
                raise ValueError('WELLPATH format error.  Need 2 or 4 tokens, got %d tokens: %s'
                                 % (num_tokens, str(wp_line)))

            depth_type = None
            depth = -1
            if num_tokens == 4:
                depth_type = tryGet( content, 'WELLPATH', i, 2)
                depth = tryGetFloat( content, 'WELLPATH', i, 3)


            date = None
            try:
                date = parse_date(date_str)
            except ValueError as err:
                logging.info(str(err))
            if date is None:
                raise ValueError('Could not read date from wellpath %d.  Got date string "%s".'
                                 % (i+1, str(date_str)))
            if depth_type:
                s.append((fname, date, depth_type, depth))
            else:
                s.append((fname, date))
        return s


def tryGetPath(cnt, key, idx1, idx2):
    try:
        return path.expanduser(cnt[key][idx1].getPath(idx2))
    except Exception as e:
        logging.warning('Failed to load path key %s from Snapwell config file.' % key)
    return None

def tryGet(cnt, key, idx1, idx2):
    try:
        return cnt[key][idx1][idx2]
    except Exception as e:
        logging.warning('Failed to load key %s from Snapwell config file.' % key)
    return None

def tryGetFloat(cnt, key, idx1, idx2):
    try:
        return float(cnt[key][idx1][idx2])
    except Exception as e:
        logging.warning('Failed to load key %s from Snapwell config file.' % key)
    return None
